### ISSUE TITLE
Crash from LazyCachedMapCollection

### ISSUE THREAD

**Issue by hyouuu** [GetStream/stream-chat-swift#2771]
## What did you do?
Nothing, saw it from crash report

## What did you expect to happen?
Not crashing

## What happened instead?
Crashed

## GetStream Environment
**GetStream Chat version:**  4.36.0
**GetStream Chat frameworks:** StreamChat, StreamChatUI
**iOS version:** 16.6
**Swift version:** 5.8
**Xcode version:** 14.3.1
**Device:** iPhone 14 Pro

## Additional context
<img width="1920" alt="Screenshot 2023-09-07 at 10 08 24 AM" src="https://github.com/GetStream/stream-chat-swift/assets/5061867/85f49d2d-da1e-4dba-9752-a8fee947ddb1">
<img width="1920" alt="Screenshot 2023-09-07 at 10 08 27 AM" src="https://github.com/GetStream/stream-chat-swift/assets/5061867/8067783b-0552-4c6d-bfd3-18cebd128347">

#0	(null) in Swift runtime failure: Unexpectedly found nil while implicitly unwrapping an Optional value ()
#1	0x00000001031c956c in specialized closure #1 in closure #1 in closure #1 in ListDatabaseObserver.startObserving() at /Users/hyouuu/Library/Developer/Xcode/DerivedData/Reclip-fggtlvneyyrqbidhfzfoiiefrrxp/SourcePackages/checkouts/stream-chat-swift/Sources/StreamChat/Controllers/ListDatabaseObserver.swift:283
#2	(null) in specialized closure #1 in closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) ()
#3	(null) in specialized closure #1 in closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) ()
#4	(null) in partial apply for specialized closure #1 in closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) ()
#5	(null) in thunk for @callee_guaranteed () -> () ()
#6	(null) in thunk for @escaping @callee_guaranteed () -> () ()
#7	(null) in developerSubmittedBlockToNSManagedObjectContextPerform ()
#8	(null) in -[NSManagedObjectContext performBlockAndWait:] ()
#9	0x000000010304c7c4 in specialized closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) at /Users/hyouuu/Library/Developer/Xcode/DerivedData/Reclip-fggtlvneyyrqbidhfzfoiiefrrxp/SourcePackages/checkouts/stream-chat-swift/Sources/StreamChat/Utils/LazyCachedMapCollection.swift:41
#10	0x0000000103155430 in LazyCachedMapCollection.subscript.getter at /Users/hyouuu/Library/Developer/Xcode/DerivedData/Reclip-fggtlvneyyrqbidhfzfoiiefrrxp/SourcePackages/checkouts/stream-chat-swift/Sources/StreamChat/Utils/LazyCachedMapCollection.swift:68
#11	(null) in LazyCachedMapCollection.subscript.read ()
#12	(null) in protocol witness for Collection.subscript.read in conformance LazyCachedMapCollection<A> ()
#13	(null) in protocol witness for IteratorProtocol.next() in conformance IndexingIterator<A> ()
#14	(null) in Zip2Sequence.Iterator.next() ()
#15	(null) in Sequence.allSatisfy(_:) ()
#16	0x0000000103155c4c in static LazyCachedMapCollection<A>.== infix(_:_:) at /Users/hyouuu/Library/Developer/Xcode/DerivedData/Reclip-fggtlvneyyrqbidhfzfoiiefrrxp/SourcePackages/checkouts/stream-chat-swift/Sources/StreamChat/Utils/LazyCachedMapCollection.swift:87
#17	(null) in AGDispatchEquatable ()
#18	(null) in AG::LayoutDescriptor::compare(unsigned char const*, unsigned char const*, unsigned char const*, unsigned long, unsigned int) ()
#19	(null) in AGGraphSetOutputValue ()
#20	(null) in closure #2 in BodyAccessor.setBody(_:) ()
#21	(null) in partial apply for closure #2 in BodyAccessor.setBody(_:) ()
#22	(null) in withUnsafePointer<A, B>(to:_:) ()
#23	(null) in ViewBodyAccessor.updateBody(of:changed:) ()
#24	(null) in DynamicBody.updateValue() ()
#25	(null) in partial apply for implicit closure #1 in closure #1 in closure #1 in Attribute.init<A>(_:) ()
#26	(null) in AG::Graph::UpdateStack::update() ()
#27	(null) in AG::Graph::update_attribute(AG::data::ptr<AG::Node>, unsigned int) ()
#28	(null) in AG::Graph::value_ref(AG::AttributeID, AGSwiftMetadata const*, unsigned char&) ()
#29	(null) in AGGraphGetValue ()
#30	(null) in GraphHost.updatePreferences() ()
#31	(null) in ViewGraph.updateOutputs() ()
#32	(null) in closure #1 in ViewRendererHost.render(interval:updateDisplayList:) ()
#33	(null) in ViewRendererHost.render(interval:updateDisplayList:) ()
#34	(null) in closure #1 in _UIHostingView.requestImmediateUpdate() ()
#35	(null) in thunk for @escaping @callee_guaranteed () -> () ()
#36	(null) in _dispatch_call_block_and_release ()
#37	(null) in _dispatch_client_callout ()
#38	(null) in _dispatch_main_queue_drain ()
#39	(null) in _dispatch_main_queue_callback_4CF ()
#40	(null) in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#41	(null) in __CFRunLoopRun ()
#42	(null) in CFRunLoopRunSpecific ()
#43	(null) in GSEventRunModal ()
#44	(null) in -[UIApplication _run] ()
#45	(null) in UIApplicationMain ()
#46	0x000000010229d3ec in main at /Users/hyouuu/reclip-ios/App/Reclip/Sources/main.swift:6
#47	(null) in start ()
Thread 1Thread 2Thread 3Thread 4Thread 5Thread 6Thread 7Thread 8Thread 9Thread 10Thread 11Thread 12Thread 13Thread 14Thread 15Thread 16Thread 17Thread 18Thread 19Thread 20Thread 21Thread 22Thread 23Thread 24Thread 25Thread 26Thread 27Thread 28Thread 29Thread 30

**Comment by nuno-vieira** [comment]
Hi @hyouuu,

Can you share a snippet of your integration when logging in / out? Usually, this issue happens with a misuse of these APIs. Often, it is because you are not waiting for the `logout()` completion block. 

Also, do you have any analytics to help understand where in the app was the user in? What screen, and what they were doing?

Best,
Nuno

**Comment by hyouuu** [comment]
Hi @nuno-vieira , we just got another crash where the user was tapping on a chat push notification to wake up the app and try to navigate to a chat. 

```
Crashed: com.apple.main-thread
0  Reclip                         0xf2b03c specialized closure #1 in closure #1 in closure #1 in ListDatabaseObserver.startObserving() + 44 (Logger.swift:44)
1  Reclip                         0xdaedb0 specialized closure #1 in closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) + 4343147952
2  Reclip                         0xdb26d8 partial apply for specialized closure #1 in closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) + 4343162584 (<compiler-generated>:4343162584)
3  Reclip                         0xf24028 thunk for @callee_guaranteed () -> () + 4344676392 (<compiler-generated>:4344676392)
4  Reclip                         0xf24048 thunk for @escaping @callee_guaranteed () -> () + 4344676424 (<compiler-generated>:4344676424)
5  CoreData                       0x80b6c developerSubmittedBlockToNSManagedObjectContextPerform + 156
6  CoreData                       0x80650 -[NSManagedObjectContext performBlockAndWait:] + 208
7  Reclip                         0xdae294 specialized closure #1 in LazyCachedMapCollection.init<A, B>(source:map:context:) + 41 (LazyCachedMapCollection.swift:41)
8  Reclip                         0xeb6f00 LazyCachedMapCollection.subscript.getter + 70 (LazyCachedMapCollection.swift:70)
9  Reclip                         0xeb72dc LazyCachedMapCollection.subscript.read + 64 (LazyCachedMapCollection.swift:64)
10 Reclip                         0xeb722c protocol witness for Collection.subscript.read in conformance LazyCachedMapCollection<A> + 4344230444 (<compiler-generated>:4344230444)
11 libswiftCore.dylib             0x78aac protocol witness for IteratorProtocol.next() in conformance IndexingIterator<A> + 552
12 libswiftCore.dylib             0x1ed1c4 Zip2Sequence.Iterator.next() + 1092
13 libswiftCore.dylib             0x1c9bd8 Sequence.allSatisfy(_:) + 648
14 Reclip                         0xeb771c static LazyCachedMapCollection<A>.== infix(_:_:) + 87 (LazyCachedMapCollection.swift:87)
15 AttributeGraph                 0x1c98 AGDispatchEquatable + 24
16 AttributeGraph                 0x3018 AG::LayoutDescriptor::compare(unsigned char const*, unsigned char const*, unsigned char const*, unsigned long, unsigned int) + 1588
17 AttributeGraph                 0x53cc AGGraphSetOutputValue + 252
18 SwiftUI                        0x2ec1c OUTLINED_FUNCTION_173 + 1816
19 SwiftUI                        0x7d54e4 OUTLINED_FUNCTION_15 + 18264
20 libswiftCore.dylib             0x173d3c withUnsafePointer<A, B>(to:_:) + 28
21 SwiftUI                        0x359a0 OUTLINED_FUNCTION_117 + 2064
22 SwiftUI                        0x53c80 OUTLINED_FUNCTION_177 + 10736
23 SwiftUI                        0x616310 OUTLINED_FUNCTION_6 + 32000
24 AttributeGraph                 0x3e48 AG::Graph::UpdateStack::update() + 520
25 AttributeGraph                 0x3748 AG::Graph::update_attribute(AG::data::ptr<AG::Node>, unsigned int) + 424
26 AttributeGraph                 0x1dd4 AG::Graph::value_ref(AG::AttributeID, AGSwiftMetadata const*, unsigned char&) + 192
27 AttributeGraph                 0x1fe0 AGGraphGetValue + 284
28 SwiftUI                        0x56758 OUTLINED_FUNCTION_209 + 820
29 SwiftUI                        0xe588 OUTLINED_FUNCTION_15 + 404
30 SwiftUI                        0x1cd54 OUTLINED_FUNCTION_157 + 2524
31 SwiftUI                        0xb7634 OUTLINED_FUNCTION_450 + 3480
32 SwiftUI                        0x811d88 OUTLINED_FUNCTION_3 + 4832
33 SwiftUI                        0x509d8 OUTLINED_FUNCTION_764 + 6480
34 libdispatch.dylib              0x2320 _dispatch_call_block_and_release + 32
35 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
36 libdispatch.dylib              0x126a4 _dispatch_main_queue_drain + 928
37 libdispatch.dylib              0x122f4 _dispatch_main_queue_callback_4CF + 44
38 CoreFoundation                 0x98c28 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
39 CoreFoundation                 0x7a560 __CFRunLoopRun + 1992
40 CoreFoundation                 0x7f3ec CFRunLoopRunSpecific + 612
41 GraphicsServices               0x135c GSEventRunModal + 164
42 UIKitCore                      0x39cf58 -[UIApplication _run] + 888
43 UIKitCore                      0x39cbbc UIApplicationMain + 340
44 Reclip                         0xc8b4 main + 6 (main.swift:6)
45 ???                            0x1c87e4dec (Missing)
```

**Comment by nuno-vieira** [comment]
Hi @hyouuu ,

Yes, you should wait for the connectUser call before any call to the Stream SDK. You also need to make sure that the connection is successful. If there is an error, you need to handle that error and display a UI to the user so that he can retry the connection. He you can find more details here: https://getstream.io/chat/docs/ios-swift/init_and_users/?language=swift#:~:text=Note%20how%20we%20are%20waiting%20for%20the%20connectUser%20API%20call%20to%20be%20completed%20before%20moving%20forward.%20You%20should%20always%20make%20sure%20to%20have%20the%20user%20set%20before%20making%20any%20more%20calls.%20All%20SDKs%20make%20this%20very%20easy%20and%20wait%20or%20queue%20requests%20until%20then.

Meanwhile, we are also investigating the issue on our side. Let us know if these changes on your side improved anything, thank you 👍 

Best,
Nuno

**Comment by nuno-vieira** [comment]
Hi @hyouuu!

Hi!

This issue should have been resolved on the new [4.37.0](https://github.com/GetStream/stream-chat-swift/releases/tag/4.37.0) Release. 

Please re-open the issue if you still have crash reports after updating.

Best,
Nuno
